catalog/files/impl/src/main/java/org/projectnessie/catalog/files/s3/S3ObjectIO.java

```diff
@@ -73,17 +73,17 @@ public InputStream readObject(URI uri) throws IOException {
   }
 
   @Override
-  public OutputStream writeObject(URI uri) {
+  public OutputStream writeObject(StorageUri uri) {
     checkArgument(uri != null, "Invalid location: null");
-    checkArgument("s3".equals(uri.getScheme()), "Invalid S3 scheme: %s", uri);
+    checkArgument("s3".equals(uri.scheme()), "Invalid S3 scheme: %s", uri);
 
     return new ByteArrayOutputStream() {
       @Override
       public void close() throws IOException {
         super.close();
 
         S3Client s3client = s3clientSupplier.getClient(uri);
-        S3Uri s3uri = s3client.utilities().parseUri(uri);
+        S3Uri s3uri = s3client.utilities().parseUri(uri.asURI());
```

Comment:
Should this be refactored to prevent using `java.net.URI` and run into the issue that this change intents to avoid?

Reply:
`uri.asURI()` produces the correct `URI` from a custom-parsed set of components (scheme, host, path, etc.).

The problem with S3 URI in Iceberg is that metadata contains location data that does not match the `URI` RFC. `StorageUri` handles that by parsing the location string directly.

Reply:
However, we cannot pass that `URI.toString()` back to Iceberg (`S3FileIO`). That `URI` works only with `S3Client`.

Reply:
Switched to using `StorageUri` components directly.... not sure now why I added the `asURI()` method initially :facepalm: